akka-cluster-sharding/src/test/scala/akka/cluster/sharding/ShardRegionSpec.scala

```diff
@@ -0,0 +1,167 @@
+
+  val storageLocation = List(
+    new File(system.settings.config.getString("akka.cluster.sharding.distributed-data.durable.lmdb.dir")).getParentFile)
+
+  // mute logging of deadLetters
+  system.eventStream.publish(Mute(DeadLettersFilter[Any]))
+
+  private val sysA = system
+  private val sysB = ActorSystem(system.name, system.settings.config)
+
+  private val p1 = TestProbe()(sysA)
+  private val p2 = TestProbe()(sysB)
+
+  private val region1 = startShard(sysA)
+  private val region2 = startShard(sysB)
+
+  override protected def atStartup(): Unit = {
+    storageLocation.foreach(dir => if (dir.exists) FileUtils.deleteQuietly(dir))
+  }
+
+  override def beforeTermination(): Unit = {
+    shutdown(sysB)
+  }
+
+  override protected def afterTermination(): Unit = {
+    storageLocation.foreach(dir => if (dir.exists) FileUtils.deleteQuietly(dir))
+  }
+
+  def startShard(sys: ActorSystem): ActorRef =
+    ClusterSharding(sys).start(
+      shardTypeName,
+      Props[EntityActor],
+      ClusterShardingSettings(system).withRememberEntities(true),
+      extractEntityId,
+      extractShardId)
+
+  def startProxy(sys: ActorSystem): ActorRef =
+    ClusterSharding(sys).startProxy(shardTypeName, None, extractEntityId, extractShardId)
+
+  "ClusterSharding" must {
+
+    "initialize cluster and allocate sharded actors" in {
+
+      Cluster(sysA).join(Cluster(sysA).selfAddress) // coordinator on A
+      awaitAssert(Cluster(sysA).selfMember.status shouldEqual MemberStatus.Up, 1.second)
+
+      Cluster(sysB).join(Cluster(sysA).selfAddress)
+
+      within(10.seconds) {
+        awaitAssert {
+          Set(sysA, sysB).foreach { s =>
+            Cluster(s).sendCurrentClusterState(testActor)
+            expectMsgType[CurrentClusterState].members.size shouldEqual 2
+          }
+        }
+      }
+
+      region1.tell(1, p1.ref)
+      p1.expectMsg(1)
+
+      region2.tell(2, p2.ref)
+      p2.expectMsg(2)
+
+      region2.tell(3, p2.ref)
+      p2.expectMsg(3)
+    }
+
+    "only deliver buffered RestartShard to the local region" in {
+
+      def statesFor(region: ActorRef, probe: TestProbe, expect: Int) = {
+        region.tell(ShardRegion.GetShardRegionState, probe.ref)
+        probe
+          .receiveWhile(messages = expect) {
+            case e: ShardRegion.CurrentShardRegionState =>
+              e.shards.map(_.shardId)
+          }
+          .flatten
+      }
+
+      def awaitRebalance(region: ActorRef, msg: Int, probe: TestProbe): Boolean = {
+        region.tell(msg, probe.ref)
+        probe.expectMsgPF(2.seconds) {
+          case id => if (id == msg) true else awaitRebalance(region, msg, probe)
+        }
+      }
+
+      val region1Shards = statesFor(region1, p1, expect = 2)
+      val region2Shards = statesFor(region2, p2, expect = 1)
+      region1Shards shouldEqual Seq("1", "3")
+      region2Shards shouldEqual Seq("2")
+      val allShards = region1Shards ++ region2Shards
+
+      region2 ! PoisonPill
```

Comment:
it could be triggered by that the `Shard` was terminated, which only realistically could happen if it is mode=persistence (which we have deprecated)

and it must also be rebalanced right after stopping the Shard, before the scheduled RestartShard message is received.

I agree that it's very difficult to create a deterministic test for this. Probably not worth spending time on it. I'm convinced that dropping the RestartShard message is the right thing to do.

Does this test add value to include if it doesn't exercise the failure scenario? Shall we just include the fix and move on?

Reply:
Hrm, originally I had the PoisonPill against the shard vs region.

I agree on value in light of the difficulty in raising this behavior to test. Though My preference is to keep it because it should prove useful in the future to easily add to. And we should be unit testing the region.